src/librustc_mir/transform/copy_prop.rs

```diff
@@ -29,18 +29,34 @@
 //! (non-mutating) use of `SRC`. These restrictions are conservative and may be relaxed in the
 //! future.
 
-use def_use::DefUseAnalysis;
-use rustc::mir::repr::{Local, Lvalue, Mir, Operand, Rvalue, StatementKind};
+use def_use::{DefUseAnalysis, MirSummary};
+use rustc::mir::repr::{Constant, Local, Location, Lvalue, Mir, Operand, Rvalue, StatementKind};
 use rustc::mir::transform::{MirPass, MirSource, Pass};
+use rustc::mir::visit::MutVisitor;
 use rustc::ty::TyCtxt;
 use rustc_data_structures::indexed_vec::Idx;
+use transform::qualify_consts;
 
 pub struct CopyPropagation;
 
 impl Pass for CopyPropagation {}
 
 impl<'tcx> MirPass<'tcx> for CopyPropagation {
-    fn run_pass<'a>(&mut self, _: TyCtxt<'a, 'tcx, 'tcx>, _: MirSource, mir: &mut Mir<'tcx>) {
+    fn run_pass<'a>(&mut self,
+                    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                    source: MirSource,
+                    mir: &mut Mir<'tcx>) {
+        match source {
+            MirSource::Const(_) | MirSource::Static(..) | MirSource::Promoted(..) => return,
+            MirSource::Fn(function_node_id) => {
+                // Don't run on const functions, because constant qualification might reject the
+                // optimized IR.
```

Comment:
Comment nit: not really a problem for const qualification (as that only looks at non-generic `const`s only, already handled by the previous match arm), but rather `rustc_trans`, which in a pre-miri world cannot evaluate MIR where mutation is used to construct ADT.